go-controller/pkg/ovn/egressfirewall_dns.go

```diff
@@ -0,0 +1,194 @@
+
+	if _, exists := e.dnsEntries[dnsName]; !exists {
+		var err error
+		dnsEntry := dnsEntry{
+			namespaces: make(map[string]struct{}),
+		}
+		if e.addressSetFactory == nil {
+			return nil, fmt.Errorf("error adding EgressFirewall DNS rule for host %s, in namespace %s: addressSetFactory is nil", dnsName, namespace)
+		}
+		dnsEntry.dnsAddressSet, err = e.addressSetFactory.NewAddressSet(dnsName, nil)
+		if err != nil {
+			return nil, fmt.Errorf("cannot create addressSet for %s: %v", dnsName, err)
+		}
+		e.dnsEntries[dnsName] = &dnsEntry
+		e.signalAdded(dnsName)
+	}
+	e.dnsEntries[dnsName].namespaces[namespace] = struct{}{}
+	return e.dnsEntries[dnsName].dnsAddressSet, nil
+
+}
+
+func (e *EgressDNS) Delete(namespace string) bool {
+	e.lock.Lock()
+	defer e.lock.Unlock()
+
+	// go through all dnsNames for namespaces
+	for dnsName, dnsEntry := range e.dnsEntries {
+		// delete the dnsEntry
+		delete(dnsEntry.namespaces, namespace)
+		if len(dnsEntry.namespaces) == 0 {
+			// the dnsEntry appears in no other namespace so delete the address_set
+			err := dnsEntry.dnsAddressSet.Destroy()
+			if err != nil {
+				klog.Errorf("Error deleteing EgressFirewall AddressSet for dnsName: %s %v", dnsName, err)
+			}
+			e.dns.Delete(dnsName)
+			// the dnsEntry is no longer needed because nothing referances it delete it
+			delete(e.dnsEntries, dnsName)
+		}
+	}
+	return len(e.dnsEntries) == 0
+}
+
+func (e *EgressDNS) Update(dns string) (bool, error) {
+	e.lock.Lock()
+	defer e.lock.Unlock()
+
+	return e.dns.Update(dns)
+}
+
+func (e *EgressDNS) updateEntryForName(dnsName string) error {
+	e.lock.Lock()
+	defer e.lock.Unlock()
+
+	ips := e.dns.GetIPs(dnsName)
+	e.dnsEntries[dnsName].dnsResolves = ips
+	if err := e.dnsEntries[dnsName].dnsAddressSet.SetIPs(ips); err != nil {
+		return fmt.Errorf("cannot add IPs from EgressFirewall AddressSet %s: %v", dnsName, err)
+	}
+	return nil
+}
+
+// Run spawns a goroutine that handles updates to the dns entries for dnsNames used in
+// EgressFirewalls. The loop runs after receiving one of two signals
+// 1. a new dnsName has been added and a signal is sent to add the new DNS name, if an
+//    EgressFirewall uses a DNS name already added by another egressFirewall the previous
+//    entry is used
+// 2. If the defaultInterval has run (30 min) without updating the DNS server is manually queried
+func (e *EgressDNS) Run(defaultInterval time.Duration) {
+	var dnsName string
+	var ttl time.Time
+	var timeSet bool
+	//intially the next DNS Query happens at the default interval
+	durationTillNextQuery := defaultInterval
+	go func() {
+		for {
+
+			// Wait for the given duration or until something gets added
+			select {
+			case dnsName := <-e.added:
+				if err := e.dns.Add(dnsName); err != nil {
+					utilruntime.HandleError(err)
+				}
+				if err := e.updateEntryForName(dnsName); err != nil {
+					utilruntime.HandleError(err)
+				}
+			case <-time.After(durationTillNextQuery):
+				if len(dnsName) > 0 {
```

Comment:
I'm kind of confused here. IIUC this EgressDNS object will handle updating for all dns entries right? If so, after the durationTillNextQuery shouldn't we iterate through all of the dnsNames that have not been updated and update them? This seems to only check the last added dnsName.

Reply:
So the GetNextQueryTime() does that for us, it goes through the map and returns the dnsName with the lowest ttl as long as it is less then the defaultInterval